proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,521 @@
+# Extend Property Wrappers to Function and Closure Parameters
+
+* Proposal: [SE-NNNN](NNNN-extend-property-wrappers-to-function-and-closure-parameters.md)
+* Authors: [Holly Borla](https://github.com/hborla), [Filip Sakel](https://github.com/filip-sakel)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#34272](https://github.com/apple/swift/pull/34272)
+
+
+## Introduction
+
+Property Wrappers were [introduced in Swift 5.1](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md), and have since become a popular feature abstracting away common accessor patterns for properties. Currently, applying a property wrapper is solely permitted at local and type context. However, with increasing adoption, demand for extending _where_ property wrappers can be applied has emerged. This proposal aims to extend property wrappers to function and closure parameters.
+
+
+## Motivation
+
+Property wrappers have undoubtably been very successful. Applying a property wrapper to a property is enabled by an incredibly lightweight and expressive syntax. Therefore, library authors can expose complex behavior through easily understandable property-wrapper types in an efficient manner. For instance, frameworks such as [SwiftUI](https://developer.apple.com/documentation/swiftui/) and [Combine](https://developer.apple.com/documentation/combine) introduce property wrappers such as [`State`](https://developer.apple.com/documentation/swiftui/state), [`Binding`](https://developer.apple.com/documentation/swiftui/binding) and [`Published`](https://developer.apple.com/documentation/combine/published) respectively, to expose elaborate behavior through a succinct interface, helping craft expressive yet simple APIs. However, property wrappers are only applicable to local and type properties, shattering the illusion that they helped realize in the first place:
+
+```swift
+@propertyWrapper
+struct Clamped<Bound : Comparable> {
+
+  init(
+    wrappedValue: Bound,
+    to range: ClosedRange<Bound>
+  ) { ... }
+  
+  
+  var wrappedValue: Bound { 
+    get { ... }
+    set { ... }
+  }
+  
+}
+
+
+struct Percentage {
+
+  @Clamped(to: 0 ... 100)
+  var percent = 0
+     
+     
+  mutating func increment() {
+    percent += 1
+    // Great!
+  }
+
+  mutating func adding(_ offset: Int) {
```

Comment:
I'm thinking we should remove this part of the example. I think it's caused a lot of people on the pitch thread to incorrectly develop a mental model for property wrapper parameters by comparing it to passing a wrapped value type. In reality, you should think of a property wrapper parameter as passing the backing wrapper type, while providing a convenient way to initialize that backing wrapper at the call-site.

Reply:
Yeah, I guess it could help clarify the mental. Not to mention that`min(100, max(0, offset))`  may confuse some people. That is, this code is essentially a clamping algorithm – which is what a proposal to add `clamping(to:)` to the standard library uses. What I mean by all this, is that people who have not used the above code to clamp a value, will most likely not understand it and, probably, get slightly confused by it.

Reply:
Yeah, absolutely. I think I'm going to change this example to use the `@Lowercased` property wrapper instead since we use that example later anyway, and I'll fix the rest in the process. Thanks!